File: src/main/kotlin/dev/jefy/connectpro/engagement/applicaion/query/ReviewQueryImpl.java
package dev.jefy.connectpro.engagement.applicaion.query;


import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import dev.jefy.connectpro.engagement.applicaion.dtos.ReviewResponse;
import dev.jefy.connectpro.engagement.domain.repository.ReviewRepository;
import dev.jefy.connectpro.engagement.domain.vo.ReviewId;
import dev.jefy.connectpro.user.UserClient;
import dev.jefy.connectpro.user.application.dtos.UserData;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */

@NullMarked
@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class ReviewQueryImpl implements ReviewQuery {
    private final UserClient userClient;
    private final ReviewRepository reviewRepo;
    @Override
    public @Nullable ReviewResponse findServiceReviewForCurrentUser(ServiceId serviceId) {
        UserData user = userClient.getCurrentUser();
        
        return reviewRepo.findById(ReviewId.of(user.id(), serviceId))
                .map(ReviewResponse::fromDomain)
                .orElse(null);
    }

    @Override
    public List<ReviewResponse> findByService(ServiceId serviceId) {
        return reviewRepo.findByIdServiceId(serviceId.value())
                .stream()
                .map(ReviewResponse::fromDomain)
                .toList();
    }
}

File: src/main/kotlin/dev/jefy/connectpro/engagement/application/query/ReviewQueryImpl.kt
package dev.jefy.connectpro.engagement.application.query

import dev.jefy.connectpro.engagement.application.dtos.ReviewResponse
import dev.jefy.connectpro.engagement.application.dtos.toResponse
import dev.jefy.connectpro.engagement.application.exception.ReviewNotFountException
import dev.jefy.connectpro.engagement.domain.repository.ReviewRepository
import dev.jefy.connectpro.engagement.domain.vo.ReviewId
import dev.jefy.connectpro.portfolio.domain.vo.ServiceId
import dev.jefy.connectpro.user.UserClient
import dev.jefy.connectpro.user.application.dtos.UserData
import dev.jefy.connectpro.user.domain.vo.UserId
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class ReviewQueryImpl(
    private val userClient: UserClient,
    private val reviewRepo: ReviewRepository
) : ReviewQuery {

    override fun findServiceReviewForCurrentUser(serviceId: ServiceId): ReviewResponse {
        val user: UserData = userClient.getCurrentUser()
        return reviewRepo.findById(ReviewId.of(UserId.of(user.id), serviceId))
            .map { it.toResponse(userClient.getData(UserId(it.id.reviewerId))) }
            .orElseThrow { ReviewNotFountException()}
    }

    override fun findByService(serviceId: ServiceId): List<ReviewResponse> {
        return reviewRepo.findByIdServiceId(serviceId.value)
            .map { it.toResponse(userClient.getData(UserId(it.id.reviewerId))) }
    }
}